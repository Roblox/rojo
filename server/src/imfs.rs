use std::{
    collections::{HashMap, HashSet},
    path::{self, Path, PathBuf},
    fmt,
    fs,
    io,
};

use failure::Fail;
use serde_derive::{Serialize, Deserialize};

use crate::project::{Project, ProjectNode};

/// A wrapper around io::Error that also attaches the path associated with the
/// error.
#[derive(Debug, Fail)]
pub struct FsError {
    #[fail(cause)]
    inner: io::Error,
    path: PathBuf,
}

impl FsError {
    fn new<P: Into<PathBuf>>(inner: io::Error, path: P) -> FsError {
        FsError {
            inner,
            path: path.into(),
        }
    }
}

impl fmt::Display for FsError {
    fn fmt(&self, output: &mut fmt::Formatter) -> fmt::Result {
        write!(output, "{}: {}", self.path.display(), self.inner)
    }
}

fn add_sync_points(imfs: &mut Imfs, project_node: &ProjectNode) -> Result<(), FsError> {
    match project_node {
        ProjectNode::Instance(node) => {
            for child in node.children.values() {
                add_sync_points(imfs, child)?;
            }
        },
        ProjectNode::SyncPoint(node) => {
            imfs.add_root(&node.path)?;
        },
    }

    Ok(())
}

/// The in-memory filesystem keeps a mirror of all files being watched by Rojo
/// in order to deduplicate file changes in the case of bidirectional syncing
/// from Roblox Studio.
///
/// It also enables Rojo to quickly generate React-like snapshots to make
/// reasoning about instances and how they relate to files easier.
#[derive(Debug, Clone)]
pub struct Imfs {
    items: HashMap<PathBuf, ImfsItem>,
    roots: HashSet<PathBuf>,
}

impl Imfs {
    pub fn new() -> Imfs {
        Imfs {
            items: HashMap::new(),
            roots: HashSet::new(),
        }
    }

    pub fn add_roots_from_project(&mut self, project: &Project) -> Result<(), FsError> {
        add_sync_points(self, &project.tree)
    }

    pub fn get_roots(&self) -> &HashSet<PathBuf> {
        &self.roots
    }

    pub fn get_items(&self) -> &HashMap<PathBuf, ImfsItem> {
        &self.items
    }

    pub fn get(&self, path: &Path) -> Option<&ImfsItem> {
        debug_assert!(path.is_absolute());
        debug_assert!(self.is_within_roots(path));

        self.items.get(path)
    }

    pub fn add_root(&mut self, path: &Path) -> Result<(), FsError> {
        debug_assert!(path.is_absolute());
        debug_assert!(!self.is_within_roots(path));

        self.roots.insert(path.to_path_buf());

        self.descend_and_read_from_disk(path)
    }

    pub fn remove_root(&mut self, path: &Path) {
        debug_assert!(path.is_absolute());

        if self.roots.get(path).is_some() {
            self.remove_item(path);

            if let Some(parent_path) = path.parent() {
                self.unlink_child(parent_path, path);
            }
        }
    }

    pub fn path_created(&mut self, path: &Path) -> Result<(), FsError> {
        debug_assert!(path.is_absolute());
        debug_assert!(self.is_within_roots(path));

        self.descend_and_read_from_disk(path)
    }

    pub fn path_updated(&mut self, path: &Path) -> Result<(), FsError> {
        debug_assert!(path.is_absolute());
        debug_assert!(self.is_within_roots(path));

        self.descend_and_read_from_disk(path)
    }

    pub fn path_removed(&mut self, path: &Path) -> Result<(), FsError> {
        debug_assert!(path.is_absolute());
        debug_assert!(self.is_within_roots(path));

        self.remove_item(path);

        if let Some(parent_path) = path.parent() {
            self.unlink_child(parent_path, path);
        }

        Ok(())
    }

    pub fn path_moved(&mut self, from_path: &Path, to_path: &Path) -> Result<(), FsError> {
        self.path_removed(from_path)?;
        self.path_created(to_path)?;
        Ok(())
    }

    pub fn get_root_for_path<'a>(&'a self, path: &Path) -> Option<&'a Path> {
        for root_path in &self.roots {
            if path.starts_with(root_path) {
                return Some(root_path);
            }
        }

        None
    }

    fn remove_item(&mut self, path: &Path) {
        if let Some(ImfsItem::Directory(directory)) = self.items.remove(path) {
            for child_path in &directory.children {
                self.remove_item(child_path);
            }
        }
    }

    fn unlink_child(&mut self, parent: &Path, child: &Path) {
        let parent_item = self.items.get_mut(parent);

        match parent_item {
            Some(ImfsItem::Directory(directory)) => {
                directory.children.remove(child);
            },
            _ => {},
        }
    }

    fn link_child(&mut self, parent: &Path, child: &Path) {
        if self.is_within_roots(parent) {
            let parent_item = self.items.get_mut(parent);

            match parent_item {
                Some(ImfsItem::Directory(directory)) => {
                    directory.children.insert(child.to_path_buf());
                },
                _ => {
                    panic!("Tried to link child of path that wasn't a directory!");
                },
            }
        }
    }

    fn descend_and_read_from_disk(&mut self, path: &Path) -> Result<(), FsError> {
        let root_path = self.get_root_path(path)
            .expect("Tried to descent and read for path that wasn't within roots!");

        // If this path is a root, we should read the entire thing.
        if root_path == path {
            self.read_from_disk(path)?;
            return Ok(());
        }

        let relative_path = path.strip_prefix(root_path).unwrap();
        let mut current_path = root_path.to_path_buf();

        for component in relative_path.components() {
            match component {
                path::Component::Normal(name) => {
                    let next_path = current_path.join(name);

                    if self.items.contains_key(&next_path) {
                        current_path = next_path;
                    } else {
                        break;
                    }
                },
                _ => unreachable!(),
            }
        }

        self.read_from_disk(&current_path)
    }

    fn read_from_disk(&mut self, path: &Path) -> Result<(), FsError> {
        let metadata = fs::metadata(path)
            .map_err(|e| FsError::new(e, path))?;

        if metadata.is_file() {
            let contents = fs::read(path)
                .map_err(|e| FsError::new(e, path))?;
            let item = ImfsItem::File(ImfsFile {
                path: path.to_path_buf(),
                contents,
            });

            self.items.insert(path.to_path_buf(), item);

            if let Some(parent_path) = path.parent() {
                self.link_child(parent_path, path);
            }

            Ok(())
        } else if metadata.is_dir() {
            let item = ImfsItem::Directory(ImfsDirectory {
                path: path.to_path_buf(),
                children: HashSet::new(),
            });

            self.items.insert(path.to_path_buf(), item);

            let dir_children = fs::read_dir(path)
                .map_err(|e| FsError::new(e, path))?;

            for entry in dir_children {
                let entry = entry
                    .map_err(|e| FsError::new(e, path))?;

                let child_path = entry.path();

                self.read_from_disk(&child_path)?;
            }

            if let Some(parent_path) = path.parent() {
                self.link_child(parent_path, path);
            }

            Ok(())
        } else {
            panic!("Unexpected non-file, non-directory item");
        }
    }

    fn get_root_path<'a>(&'a self, path: &Path) -> Option<&'a Path> {
        for root_path in &self.roots {
            if path.starts_with(root_path) {
                return Some(root_path)
            }
        }

        None
    }

    fn is_within_roots(&self, path: &Path) -> bool {
        for root_path in &self.roots {
            if path.starts_with(root_path) {
                return true;
            }
        }

        false
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ImfsFile {
    pub path: PathBuf,
    pub contents: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ImfsDirectory {
    pub path: PathBuf,
    pub children: HashSet<PathBuf>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum ImfsItem {
    File(ImfsFile),
    Directory(ImfsDirectory),
}