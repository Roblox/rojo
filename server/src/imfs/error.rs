use std::{
    io,
    fmt,
    path::PathBuf,
};

use failure::Fail;

pub type FsResult<T> = Result<T, FsError>;
pub use io::ErrorKind as FsErrorKind;

pub trait FsResultExt<T> {
    fn with_not_found(self) -> Result<Option<T>, FsError>;
}

impl<T> FsResultExt<T> for Result<T, FsError> {
    fn with_not_found(self) -> Result<Option<T>, FsError> {
        match self {
            Ok(value) => Ok(Some(value)),
            Err(ref err) if err.kind() == FsErrorKind::NotFound => Ok(None),
            Err(err) => Err(err),
        }
    }
}

// TODO: New error type that contains errors specific to our application,
// wrapping io::Error either directly or through another error type that has
// path information.
//
// It's possible that we should hoist up the path information one more level, or
// destructure/restructure information to hoist the path out of FsError and just
// embed io::Error?
pub enum ImfsError {
    NotFound,
    WrongKind,
    Io(io::Error),
}

/// A wrapper around io::Error that also attaches the path associated with the
/// error.
#[derive(Debug, Fail)]
pub struct FsError {
    #[fail(cause)]
    inner: io::Error,
    path: PathBuf,
}

impl FsError {
    pub fn new<P: Into<PathBuf>>(inner: io::Error, path: P) -> FsError {
        FsError {
            inner,
            path: path.into(),
        }
    }

    pub fn kind(&self) -> FsErrorKind {
        self.inner.kind()
    }
}

impl fmt::Display for FsError {
    fn fmt(&self, output: &mut fmt::Formatter) -> fmt::Result {
        write!(output, "{}: {}", self.path.display(), self.inner)
    }
}