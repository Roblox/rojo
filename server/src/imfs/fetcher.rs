use std::{
    io,
    path::{Path, PathBuf},
};

use crossbeam_channel::Receiver;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FileType {
    File,
    Directory,
}

// TODO: Use our own event type instead of notify's.
pub type ImfsEvent = notify::DebouncedEvent;

/// The generic interface that `Imfs` uses to lazily read files from the disk.
/// In tests, it's stubbed out to do different versions of absolutely nothing
/// depending on the test.
pub trait ImfsFetcher {
    fn file_type(&mut self, path: &Path) -> io::Result<FileType>;
    fn read_children(&mut self, path: &Path) -> io::Result<Vec<PathBuf>>;
    fn read_contents(&mut self, path: &Path) -> io::Result<Vec<u8>>;

    fn create_directory(&mut self, path: &Path) -> io::Result<()>;
    fn write_file(&mut self, path: &Path, contents: &[u8]) -> io::Result<()>;
    fn remove(&mut self, path: &Path) -> io::Result<()>;

    fn watch(&mut self, path: &Path);
    fn unwatch(&mut self, path: &Path);
    fn receiver(&mut self) -> Receiver<ImfsEvent>;
}