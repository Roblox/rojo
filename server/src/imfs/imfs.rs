use std::{
    io,
    path::{Path, PathBuf},
};

use crate::path_map::PathMap;

use super::{
    snapshot::ImfsSnapshot,
    error::{FsResult, FsError},
    fetcher::{ImfsFetcher, FileType},
};

/// An in-memory filesystem that can be incrementally populated and updated as
/// filesystem modification events occur.
///
/// All operations on the `Imfs` are lazy and do I/O as late as they can to
/// avoid reading extraneous files or directories from the disk. This means that
/// they all take `self` mutably, and means that it isn't possible to hold
/// references to the internal state of the Imfs while traversing it!
///
/// Most operations return `ImfsEntry` objects to work around this, which is
/// effectively a index into the `Imfs`.
pub struct Imfs<F> {
    inner: PathMap<ImfsItem>,
    fetcher: F,
}

impl<F: ImfsFetcher> Imfs<F> {
    pub fn new(fetcher: F) -> Imfs<F> {
        Imfs {
            inner: PathMap::new(),
            fetcher,
        }
    }

    pub fn load_from_snapshot(&mut self, path: impl AsRef<Path>, snapshot: ImfsSnapshot) {
        let path = path.as_ref();

        match snapshot {
            ImfsSnapshot::File(file) => {
                self.inner.insert(path.to_path_buf(), ImfsItem::File(ImfsFile {
                    path: path.to_path_buf(),
                    contents: Some(file.contents),
                }));
            }
            ImfsSnapshot::Directory(directory) => {
                self.inner.insert(path.to_path_buf(), ImfsItem::Directory(ImfsDirectory {
                    path: path.to_path_buf(),
                    children_enumerated: true,
                }));

                for (child_name, child) in directory.children.into_iter() {
                    self.load_from_snapshot(path.join(child_name), child);
                }
            }
        }
    }

    pub fn raise_file_change(&mut self, path: impl AsRef<Path>) -> FsResult<()> {
        let path = path.as_ref();

        if !self.would_be_resident(path) {
            return Ok(());
        }

        let new_type = self.fetcher.file_type(path)
            .map_err(|err| FsError::new(err, path.to_path_buf()))?;

        match self.inner.get_mut(path) {
            Some(existing_item) => {
                match (existing_item, &new_type) {
                    (ImfsItem::File(existing_file), FileType::File) => {
                        // Invalidate the existing file contents.
                        // We can probably be smarter about this by reading the changed file.
                        existing_file.contents = None;
                    }
                    (ImfsItem::Directory(_), FileType::Directory) => {
                        // No changes required, a directory updating doesn't mean anything to us.
                    }
                    (ImfsItem::File(_), FileType::Directory) => {
                        self.inner.remove(path);
                        self.inner.insert(path.to_path_buf(), ImfsItem::new_from_type(FileType::Directory, path));
                    }
                    (ImfsItem::Directory(_), FileType::File) => {
                        self.inner.remove(path);
                        self.inner.insert(path.to_path_buf(), ImfsItem::new_from_type(FileType::File, path));
                    }
                }
            }
            None => {
                self.inner.insert(path.to_path_buf(), ImfsItem::new_from_type(new_type, path));
            }
        }

        Ok(())
    }

    pub fn raise_file_removed(&mut self, path: impl AsRef<Path>) -> FsResult<()> {
        let path = path.as_ref();

        if !self.would_be_resident(path) {
            return Ok(());
        }

        self.inner.remove(path);
        Ok(())
    }

    pub fn get(&mut self, path: impl AsRef<Path>) -> FsResult<ImfsEntry> {
        self.read_if_not_exists(path.as_ref())?;

        let item = self.inner.get(path.as_ref()).unwrap();

        let is_file = match item {
            ImfsItem::File(_) => true,
            ImfsItem::Directory(_) => false,
        };

        Ok(ImfsEntry {
            path: item.path().to_path_buf(),
            is_file,
        })
    }

    pub fn get_contents(&mut self, path: impl AsRef<Path>) -> FsResult<&[u8]> {
        let path = path.as_ref();

        self.read_if_not_exists(path)?;

        match self.inner.get_mut(path).unwrap() {
            ImfsItem::File(file) => {
                if file.contents.is_none() {
                    file.contents = Some(self.fetcher.read_contents(path)
                        .map_err(|err| FsError::new(err, path.to_path_buf()))?);
                }

                Ok(file.contents.as_ref().unwrap())
            }
            ImfsItem::Directory(_) => Err(FsError::new(io::Error::new(io::ErrorKind::Other, "Can't read a directory"), path.to_path_buf()))
        }
    }

    pub fn get_children(&mut self, path: impl AsRef<Path>) -> FsResult<Vec<ImfsEntry>> {
        let path = path.as_ref();

        self.read_if_not_exists(path)?;

        match self.inner.get(path).unwrap() {
            ImfsItem::Directory(dir) => {
                if dir.children_enumerated {
                    return self.inner.children(path)
                        .unwrap() // TODO: Handle None here, which means the PathMap entry did not exist.
                        .into_iter()
                        .map(PathBuf::from) // Convert paths from &Path to PathBuf
                        .collect::<Vec<PathBuf>>() // Collect all PathBufs, since self.get needs to borrow self mutably.
                        .into_iter()
                        .map(|path| self.get(path))
                        .collect::<FsResult<Vec<ImfsEntry>>>();
                }

                self.fetcher.read_children(path)
                    .map_err(|err| FsError::new(err, path.to_path_buf()))?
                    .into_iter()
                    .map(|path| self.get(path))
                    .collect::<FsResult<Vec<ImfsEntry>>>()
            }
            ImfsItem::File(_) => Err(FsError::new(io::Error::new(io::ErrorKind::Other, "Can't read a directory"), path.to_path_buf()))
        }
    }

    /// Tells whether the given path, if it were loaded, would be loaded if it
    /// existed.
    ///
    /// Returns true if the path is loaded or if its parent is loaded, is a
    /// directory, and is marked as having been enumerated before.
    ///
    /// This idea corresponds to whether a file change event should result in
    /// tangible changes to the in-memory filesystem. If a path would be
    /// resident, we need to read it, and if its contents were known before, we
    /// need to update them.
    fn would_be_resident(&self, path: &Path) -> bool {
        if self.inner.contains_key(path) {
            return true;
        }

        if let Some(parent) = path.parent() {
            if let Some(ImfsItem::Directory(dir)) = self.inner.get(parent) {
                return !dir.children_enumerated;
            }
        }

        false
    }

    /// Attempts to read the path into the `Imfs` if it doesn't exist.
    ///
    /// This does not necessitate that file contents or directory children will
    /// be read. Depending on the `ImfsFetcher` implementation that the `Imfs`
    /// is using, this call may read exactly only the given path and no more.
    fn read_if_not_exists(&mut self, path: &Path) -> FsResult<()> {
        if !self.inner.contains_key(path) {
            let kind = self.fetcher.file_type(path)
                .map_err(|err| FsError::new(err, path.to_path_buf()))?;

            self.inner.insert(path.to_path_buf(), ImfsItem::new_from_type(kind, path));
        }

        Ok(())
    }
}

/// A reference to file or folder in an `Imfs`. Can only be produced by the
/// entry existing in the Imfs, but can later point to nothing if something
/// would invalidate that path.
///
/// This struct does not borrow from the Imfs since every operation has the
/// possibility to mutate the underlying data structure and move memory around.
pub struct ImfsEntry {
    path: PathBuf,
    is_file: bool,
}

impl ImfsEntry {
    pub fn path(&self) -> &Path {
        &self.path
    }

    pub fn contents<'imfs>(
        &self,
        imfs: &'imfs mut Imfs<impl ImfsFetcher>,
    ) -> FsResult<&'imfs [u8]> {
        imfs.get_contents(&self.path)
    }

    pub fn children(
        &self,
        imfs: &mut Imfs<impl ImfsFetcher>,
    ) -> FsResult<Vec<ImfsEntry>> {
        imfs.get_children(&self.path)
    }

    pub fn is_file(&self) -> bool {
        self.is_file
    }

    pub fn is_directory(&self) -> bool {
        !self.is_file
    }
}

/// Internal structure describing potentially partially-resident files and
/// folders in the `Imfs`.
pub enum ImfsItem {
    File(ImfsFile),
    Directory(ImfsDirectory),
}

impl ImfsItem {
    fn path(&self) -> &Path {
        match self {
            ImfsItem::File(file) => &file.path,
            ImfsItem::Directory(dir) => &dir.path,
        }
    }

    fn new_from_type(kind: FileType, path: impl Into<PathBuf>) -> ImfsItem {
        match kind {
            FileType::Directory => ImfsItem::Directory(ImfsDirectory {
                path: path.into(),
                children_enumerated: false,
            }),
            FileType::File => ImfsItem::File(ImfsFile {
                path: path.into(),
                contents: None,
            }),
        }
    }
}

pub struct ImfsFile {
    pub(super) path: PathBuf,
    pub(super) contents: Option<Vec<u8>>,
}

pub struct ImfsDirectory {
    pub(super) path: PathBuf,
    pub(super) children_enumerated: bool,
}

#[cfg(test)]
mod test {
    use super::*;

    use std::{
        rc::Rc,
        cell::RefCell,
    };

    use maplit::hashmap;

    use super::super::{
        noop_fetcher::NoopFetcher,
        error::FsErrorKind,
    };

    #[test]
    fn from_snapshot_file() {
        let mut imfs = Imfs::new(NoopFetcher);
        let file = ImfsSnapshot::file("hello, world!");

        imfs.load_from_snapshot("/hello.txt", file);

        let entry = imfs.get_contents("/hello.txt").unwrap();
        assert_eq!(entry, b"hello, world!");
    }

    #[test]
    fn from_snapshot_dir() {
        let mut imfs = Imfs::new(NoopFetcher);
        let dir = ImfsSnapshot::dir(hashmap! {
            "a.txt" => ImfsSnapshot::file("contents of a.txt"),
            "b.lua" => ImfsSnapshot::file("contents of b.lua"),
        });

        imfs.load_from_snapshot("/dir", dir);

        let children = imfs.get_children("/dir").unwrap();

        let mut has_a = false;
        let mut has_b = false;

        for child in children.into_iter() {
            if child.path() == Path::new("/dir/a.txt") {
                has_a = true;
            } else if child.path() == Path::new("/dir/b.lua") {
                has_b = true;
            } else {
                panic!("Unexpected child in /dir");
            }
        }

        assert!(has_a, "/dir/a.txt was missing");
        assert!(has_b, "/dir/b.lua was missing");

        let a = imfs.get_contents("/dir/a.txt").unwrap();
        assert_eq!(a, b"contents of a.txt");

        let b = imfs.get_contents("/dir/b.lua").unwrap();
        assert_eq!(b, b"contents of b.lua");
    }

    #[test]
    fn changed_event() {
        #[derive(Default)]
        struct MockState {
            a_contents: &'static str,
        }

        struct MockFetcher {
            inner: Rc<RefCell<MockState>>,
        }

        impl ImfsFetcher for MockFetcher {
            fn file_type(&mut self, path: &Path) -> io::Result<FileType> {
                if path == Path::new("/dir/a.txt") {
                    return Ok(FileType::File);
                }

                unimplemented!();
            }

            fn read_contents(&mut self, path: &Path) -> io::Result<Vec<u8>> {
                if path == Path::new("/dir/a.txt") {
                    let inner = self.inner.borrow();

                    return Ok(Vec::from(inner.a_contents));
                }

                unimplemented!();
            }

            fn read_children(&mut self, _path: &Path) -> io::Result<Vec<PathBuf>> {
                unimplemented!();
            }

            fn create_directory(&mut self, _path: &Path) -> io::Result<()> {
                unimplemented!();
            }

            fn write_file(&mut self, _path: &Path, _contents: &[u8]) -> io::Result<()> {
                unimplemented!();
            }

            fn remove(&mut self, _path: &Path) -> io::Result<()> {
                unimplemented!();
            }
        }

        let mock_state = Rc::new(RefCell::new(MockState {
            a_contents: "Initial contents",
        }));

        let mut imfs = Imfs::new(MockFetcher {
            inner: mock_state.clone(),
        });

        let a = imfs.get("/dir/a.txt")
            .expect("mock file did not exist");

        let contents = a.contents(&mut imfs)
            .expect("mock file contents error");

        assert_eq!(contents, b"Initial contents");

        {
            let mut mock_state = mock_state.borrow_mut();
            mock_state.a_contents = "Changed contents";
        }

        imfs.raise_file_change("/dir/a.txt")
            .expect("error processing file change");

        let contents = a.contents(&mut imfs)
            .expect("mock file contents error");

        assert_eq!(contents, b"Changed contents");
    }

    #[test]
    fn removed_event_existing() {
        let mut imfs = Imfs::new(NoopFetcher);

        let file = ImfsSnapshot::file("hello, world!");
        imfs.load_from_snapshot("/hello.txt", file);

        let hello = imfs.get("/hello.txt")
            .expect("couldn't get hello.txt");

        let contents = hello.contents(&mut imfs)
            .expect("couldn't get hello.txt contents");

        assert_eq!(contents, b"hello, world!");

        imfs.raise_file_removed("/hello.txt")
            .expect("error processing file removal");

        match imfs.get("hello.txt") {
            Err(ref err) if err.kind() == FsErrorKind::NotFound => {}
            Ok(_) => {
                panic!("hello.txt was not removed from Imfs");
            }
            Err(err) => {
                panic!("Unexpected error: {:?}", err);
            }
        }
    }
}