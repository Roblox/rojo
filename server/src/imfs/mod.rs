mod error;
mod fetcher;
mod imfs;
mod legacy;
mod noop_fetcher;
mod real_fetcher;
mod snapshot;
mod watcher;

pub use legacy::*;
pub use error::*;

pub mod new {
    pub use super::error::*;
    pub use super::imfs::*;
    pub use super::fetcher::*;
    pub use super::real_fetcher::*;
    pub use super::noop_fetcher::*;
    pub use super::snapshot::*;
    pub use super::watcher::*;
}