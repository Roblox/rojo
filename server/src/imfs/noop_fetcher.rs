//! Implements the IMFS fetcher interface for a fake filesystem using Rust's
//! std::fs interface.

use std::{
    io,
    path::{Path, PathBuf},
};

use super::{
    imfs::ImfsItem,
    fetcher::ImfsFetcher,
};

pub struct NoopFetcher;

impl ImfsFetcher for NoopFetcher {
    fn read_item(&mut self, _path: &Path) -> io::Result<ImfsItem> {
        Err(io::Error::new(io::ErrorKind::NotFound, "no-op"))
    }

    fn read_children(&mut self, _path: &Path) -> io::Result<Vec<PathBuf>> {
        Err(io::Error::new(io::ErrorKind::NotFound, "no-op"))
    }

    fn read_contents(&mut self, _path: &Path) -> io::Result<Vec<u8>> {
        Err(io::Error::new(io::ErrorKind::NotFound, "no-op"))
    }

    fn create_directory(&mut self, _path: &Path) -> io::Result<()> {
        Ok(())
    }

    fn write_file(&mut self, _path: &Path, _contents: &[u8]) -> io::Result<()> {
        Ok(())
    }

    fn remove(&mut self, _path: &Path) -> io::Result<()> {
        Ok(())
    }
}