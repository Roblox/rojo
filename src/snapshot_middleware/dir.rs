use std::{borrow::Cow, collections::HashMap};

use rbx_dom_weak::{RbxId, RbxTree};

use crate::{
    imfs::{DirectorySnapshot, Imfs, ImfsEntry, ImfsFetcher, ImfsSnapshot},
    snapshot::{InstanceMetadata, InstanceSnapshot},
};

use super::{
    context::InstanceSnapshotContext,
    error::SnapshotError,
    middleware::{SnapshotFileResult, SnapshotInstanceResult, SnapshotMiddleware},
    snapshot_from_imfs, snapshot_from_instance,
};

pub struct SnapshotDir;

impl SnapshotMiddleware for SnapshotDir {
    fn from_imfs<F: ImfsFetcher>(
        context: &mut InstanceSnapshotContext,
        imfs: &mut Imfs<F>,
        entry: &ImfsEntry,
    ) -> SnapshotInstanceResult<'static> {
        if entry.is_file() {
            return Ok(None);
        }

        let children: Vec<ImfsEntry> = entry.children(imfs)?;

        let mut snapshot_children = Vec::new();

        for child in children.into_iter() {
            if let Some(child_snapshot) = snapshot_from_imfs(context, imfs, &child)? {
                snapshot_children.push(child_snapshot);
            }
        }

        let instance_name = entry
            .path()
            .file_name()
            .expect("Could not extract file name")
            .to_str()
            .ok_or_else(|| SnapshotError::file_name_bad_unicode(entry.path()))?
            .to_string();

        Ok(Some(InstanceSnapshot {
            snapshot_id: None,
            metadata: InstanceMetadata {
                instigating_source: Some(entry.path().to_path_buf().into()),
                relevant_paths: vec![entry.path().to_path_buf()],
                ..Default::default()
            },
            name: Cow::Owned(instance_name),
            class_name: Cow::Borrowed("Folder"),
            properties: HashMap::new(),
            children: snapshot_children,
        }))
    }

    fn from_instance(tree: &RbxTree, id: RbxId) -> SnapshotFileResult {
        let instance = tree.get_instance(id).unwrap();

        if instance.class_name != "Folder" {
            return None;
        }

        let mut children = HashMap::new();

        for child_id in instance.get_children_ids() {
            if let Some((name, child)) = snapshot_from_instance(tree, *child_id) {
                children.insert(name, child);
            }
        }

        let snapshot = ImfsSnapshot::Directory(DirectorySnapshot { children });

        Some((instance.name.clone(), snapshot))
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_yaml_snapshot;
    use maplit::hashmap;

    use crate::imfs::{ImfsDebug, NoopFetcher};

    #[test]
    fn empty_folder() {
        let mut imfs = Imfs::new(NoopFetcher);
        let dir = ImfsSnapshot::dir::<String>(HashMap::new());

        imfs.debug_load_snapshot("/foo", dir);

        let entry = imfs.get("/foo").unwrap();
        let instance_snapshot =
            SnapshotDir::from_imfs(&mut InstanceSnapshotContext::default(), &mut imfs, &entry)
                .unwrap()
                .unwrap();

        assert_yaml_snapshot!(instance_snapshot);
    }

    #[test]
    fn folder_in_folder() {
        let mut imfs = Imfs::new(NoopFetcher);
        let dir = ImfsSnapshot::dir(hashmap! {
            "Child" => ImfsSnapshot::dir::<String>(HashMap::new()),
        });

        imfs.debug_load_snapshot("/foo", dir);

        let entry = imfs.get("/foo").unwrap();
        let instance_snapshot =
            SnapshotDir::from_imfs(&mut InstanceSnapshotContext::default(), &mut imfs, &entry)
                .unwrap()
                .unwrap();

        assert_yaml_snapshot!(instance_snapshot);
    }
}
