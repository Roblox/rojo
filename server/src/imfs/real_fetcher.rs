//! Implements the IMFS fetcher interface for the real filesystem using Rust's
//! std::fs interface and notify as the file watcher.

use std::{
    fs,
    io,
    path::{Path, PathBuf},
    sync::mpsc,
    time::Duration,
};

use jod_thread::JoinHandle;
use crossbeam_channel::{Receiver, unbounded};
use notify::{RecursiveMode, RecommendedWatcher, Watcher};

use super::fetcher::{ImfsFetcher, FileType, ImfsEvent};

pub struct RealFetcher {
    watcher: RecommendedWatcher,

    /// Thread handle to convert notify's mpsc channel messages into
    /// crossbeam_channel messages.
    _converter_thread: JoinHandle<()>,
    receiver: Receiver<ImfsEvent>,
}

impl RealFetcher {
    pub fn new() -> RealFetcher {
        let (notify_sender, notify_receiver) = mpsc::channel();
        let (s, r) = unbounded();

        let watcher = notify::watcher(notify_sender, Duration::from_millis(300))
            .expect("Couldn't start 'notify' file watcher");

        let handle = jod_thread::Builder::new()
            .name("RealFetcher message converter".to_owned())
            .spawn(move || {
                log::trace!("Starting RealFetcher message converter thread");

                notify_receiver.into_iter()
                    .for_each(|event| { s.send(event).unwrap() });
            })
            .expect("Could not start message converter thread");

        RealFetcher {
            watcher,
            _converter_thread: handle,
            receiver: r,
        }
    }
}

impl ImfsFetcher for RealFetcher {
    fn file_type(&mut self, path: &Path) -> io::Result<FileType> {
        let metadata = fs::metadata(path)?;

        if metadata.is_file() {
            Ok(FileType::File)
        } else {
            Ok(FileType::Directory)
        }
    }

    fn read_children(&mut self, path: &Path) -> io::Result<Vec<PathBuf>> {
        log::trace!("Reading directory {}", path.display());

        let mut result = Vec::new();

        let iter = fs::read_dir(path)?;

        for entry in iter {
            result.push(entry?.path());
        }

        Ok(result)
    }

    fn read_contents(&mut self, path: &Path) -> io::Result<Vec<u8>> {
        log::trace!("Reading file {}", path.display());

        fs::read(path)
    }

    fn create_directory(&mut self, path: &Path) -> io::Result<()> {
        log::trace!("Creating directory {}", path.display());

        fs::create_dir(path)
    }

    fn write_file(&mut self, path: &Path, contents: &[u8]) -> io::Result<()> {
        log::trace!("Writing path {}", path.display());

        fs::write(path, contents)
    }

    fn remove(&mut self, path: &Path) -> io::Result<()> {
        log::trace!("Removing path {}", path.display());

        let metadata = fs::metadata(path)?;

        if metadata.is_file() {
            fs::remove_file(path)
        } else {
            fs::remove_dir_all(path)
        }
    }

    fn watch(&mut self, path: &Path) {
        log::trace!("Watching path {}", path.display());

        if let Err(err) = self.watcher.watch(path, RecursiveMode::NonRecursive) {
            log::warn!("Couldn't watch path {}: {:?}", path.display(), err);
        }
    }

    fn unwatch(&mut self, path: &Path) {
        log::trace!("Stopped watching path {}", path.display());

        if let Err(err) = self.watcher.unwatch(path) {
            log::warn!("Couldn't unwatch path {}: {:?}", path.display(), err);
        }
    }

    fn receiver(&self) -> Receiver<ImfsEvent> {
        self.receiver.clone()
    }
}