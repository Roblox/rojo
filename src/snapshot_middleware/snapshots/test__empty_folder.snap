---
source: src/snapshot_middleware/dir.rs
expression: instance_snapshot
---
snapshot_id: ~
metadata:
  ignore_unknown_instances: false
  contributing_paths:
    - /foo
name: foo
class_name: Folder
properties: {}
children: []
